Add Interactive Web Tour for Virtual Studies and Group Comparison

Beking0912 opened the pull request:
## Summarize
- Student Name: [Biqing (Gloria) Su](https://github.com/Beking0912)
- Mentors: [Jeremy R. Easton-Marks](https://github.com/JREastonMarks), [Ryan Fu](https://github.com/fuzhaoyuan)
- Organization: [cBioportal for Cancer Genomics](https://www.cbioportal.org/)
- GSoC Project: [Interactive web tour](https://summerofcode.withgoogle.com/programs/2023/projects/luumPTO5)
- Related Issue: https://github.com/cBioPortal/GSoC/issues/86.
- E2E Testing PR: https://github.com/cBioPortal/cbioportal-frontend/pull/4700
- Related Documentation: [documentation](https://docs.google.com/document/d/1Xj8ZfcWcv2n_stCM26hMgk63xnxq7s7_W5RUYwBNUoQ/edit#heading=h.f5bc9dunvzs3).
- Online review: https://deploy-preview-4687--cbioportalfrontend.netlify.app/

## Description
This pull request introduces an Interactive Web Tour that aims to enhance the user experience and facilitate the exploration of less prominent features on the site. The interactive web tour guides new users through the functionalities of Virtual Studies (**Create a Virtual Study - Not Logged In and Logged In**) as well as Group Comparison (**Compare User-defined Groups of Samples**). The project is one of this year's Google Summer of Code (GSoC) projects.

The tour is implemented using the [reactour](https://github.com/elrumordelaluz/reactour/tree/v1) library to provide an interactive UI, leveraging localStorage for seamless page transitions and TypeScript for robust type definitions. 

## Related Issue
- e2e testing PR: https://github.com/cBioPortal/cbioportal-frontend/pull/4700
- Add `show_web_tours` for Interactive Web Tour PR: https://github.com/cBioPortal/cbioportal/pull/10333
- GSoC project description: https://github.com/cBioPortal/GSoC/issues/86

## Achievements

- Implemented an extensible tour component, allowing for easy addition of new tour by creating new step file.
- Created tours for two major features:
   - Virtual Studies
      - Create a Virtual Study for Not Logged In users
      - Create a Virtual Study for Logged In users
   - Group Comparison
      - Compare User-defined Groups of Samples
- Completed end-to-end tests for the above two tour functionalities. 


## File Structure

- `src/tours`
  - `Steps`
    - `GroupComparison.tsx`   // Steps for group comparison 
    - `VirtualStudy.tsx`   // Steps for virtual study 
    - `index.tsx`   // Exports all the steps under the folder 
  - `Tour`
    - `index.tsx`   // Main file, the entry for tours and the tour that actually runs 
    - `types.tsx`   // Types for functions and props
    - `styles.scss`
  - `index.tsx`   // Exports all tours here

## Testing
E2E tests have been completed for the Create a Virtual Study for Logged In users and Compare User-defined Groups of Samples tour. The tests focus on validating the presence and functionality of expected buttons and elements by simulating user interactions. (Note: Test code is not included in this pull request but will be added separately.)
👉 **See more about e2e testing on PR https://github.com/cBioPortal/cbioportal-frontend/pull/4700.**

## Documentation
For detailed steps of each tour, test procedures, and instructions on adding new tours, refer to the [detailed documentation](https://docs.google.com/document/d/1Xj8ZfcWcv2n_stCM26hMgk63xnxq7s7_W5RUYwBNUoQ/edit#heading=h.f5bc9dunvzs3).

## Demonstration
- Create a Virtual Study for Not Logged In users: [video](https://drive.google.com/file/d/175nxZWAI6kkPxEmLnhrfHenKUbVj7M90/view?usp=sharing)
- Create a Virtual Study for Logged In users: [video](https://drive.google.com/file/d/1BX8Ra42a5h9aPvfby-NJuyRvIK6jgRAX/view?usp=drive_link)
- Compare User-defined Groups of Samples: [video](https://drive.google.com/file/d/1IL1a67X8SiZ_1-VzzPa9QGlO3IHhLGfs/view?usp=drive_link)

## UI Screenshots
Tour Entries in on the right bar:
<img src="https://s2.loli.net/2023/08/08/SILc6uUJhVmeoTO.png" alt="Tour Entries" width="50%"/>

Normal Step UI:
<img src="https://s2.loli.net/2023/08/08/sEXbSf3ylHjwIut.png" alt="Normal Step" width="65%"/>

Last Step UI:
<img src="https://s2.loli.net/2023/08/08/mCT7tK9GZh8SJF5.png" alt="Last Step" width="65%"/>

inodb commented:
This looks really great!

Quick comments:

- [ ] 
![image](https://github.com/cBioPortal/cbioportal-frontend/assets/1334004/d851ea29-9597-4306-be1d-e108fa78b4b9)
 "Click the check box in the “#” column to select three samples such as IDH1 mutations, TP53 mutant samples and EGFR amplified samples." -> Click the check box in the “#” column to select samples with e.g. IDH1 or TP53 mutations.
- [ ] In that same step, when you select two and then deselect one, I see this screen:
![image](https://github.com/cBioPortal/cbioportal-frontend/assets/1334004/8c6fe514-be5a-461d-b62d-f71efbd9833c)

Beking0912 commented:
> 

Thanks for your feedback on the PR, Ino. I've fixed these two issues in the commit a28c584.
1. I modified the text at that step:
**From:** Select three samples
**To:** Select two samples
**From:** Click the check box in the “#” column to select three samples such as IDH1 mutations, TP53 mutant samples and EGFR amplified samples.
**To:** Click the check box in the “#” column to select samples with e.g. IDH1 or TP53 mutations.
2. Now, when the user selects two "#", the process advances automatically to the next step without the option to deselect.

inodb commented:
Great! Thanks so much @Beking0912 

Some other thoughts, which we can prolly follow up on later. Prolly not necessary for this PR, but just writing it down:

- When you click anywhere else on the page one exists the tour. This is great but sometimes one might accidentally click in the wrong area (happened to me :). Not sure if there's an easy way to bring the tour back. Alternatively maybe an additional confirmation button could work ("Do you really want to exit the tour?")
- Is there a way to enforce selection of a specific cohort? It might make sense to highlight some more specific cancer biology in future tours. Like e.g. in the mutations table one can explain what the most commonly mutated gene in glioma is. For the genomic alterations one could use e.g. [this example](https://www.cbioportal.org/comparison/alterations?sessionId=5cf89323e4b0ab413787436c&selectedGene=AR) and explain why AR is enriched in metastatic samples

inodb commented on src/shared/components/rightbar/RightBar.tsx:
```diff
@@ -303,9 +304,31 @@ export default class RightBar extends React.Component<
+            !this.studyStore.cancerTypes.isComplete ||
+            !this.studyStore.cancerStudies.isComplete ||
+            !this.studyStore.userVirtualStudies.isComplete;
+        return getServerConfig().skin_right_nav_show_testimonials ? (
```
maybe we can add another property for tours specifically?

Beking0912 replied:
Thanks for reminder!! I'll add one.

inodb commented:
@Beking0912 apologies, one more request from the product team (Niki). On the frontpage could you change:

- [ ] the tour section to be below the "What's New?" section?
- [ ] Rename the "Web tours" header to "Interactive Tours"

alisman commented on src/pages/studyView/StudyViewPage.tsx:
```diff
@@ -1016,6 +1017,15 @@ export default class StudyViewPage extends React.Component<
     }
 
     render() {
+        const isLoading =
```
lets put these in a computed getter of the component

alisman commented on src/pages/studyView/StudyViewPage.tsx:
```diff
@@ -1016,6 +1017,15 @@ export default class StudyViewPage extends React.Component<
+            this.store.queriedSampleIdentifiers.isPending ||
+            this.store.invalidSampleIds.isPending ||
+            this.body.isPending;
+        const studies =
```
we need to check that these are not pending.

Beking0912 replied:
Thanks for your feedback！I apologize for the confusion caused by the naming. Here, I only need to determine if any samples were selected, without needing the specific count.

Beking0912 commented:
Thanks for your feedback on the PR, @inodb @alisman . I'd like to share some updates:

- **DONE** the tour section to be below the "What's New?" section. 
- **DONE** Rename the "Web tours" header to "Interactive Tours".
- **DONE** put `isLoading` in a computed getter of the component.
- **DONE** Rename the function that determines number of studies.
- **DONE** Add the property `show_web_tours` to hide/show the web tour on frontend. (I'll create another PR on backend repo for adding `show_web_tours`, will update the PR link here later)

Feel free to give more feedback!

alisman commented on src/tours/Tour/styles.scss:
```diff
@@ -0,0 +1,55 @@
+.interactive-tour {
```
if possible we should use css modules for this.  it may be precluded by use of react-tour library but we can discuss

alisman commented on src/tours/Tour/index.tsx:
```diff
@@ -0,0 +1,140 @@
+    return (
+        <div>
+            {!hideEntry &&
+                Object.keys(toursMap).map(tourType => {
```
use lodash for slightly better syntax:
```
_.forIn(toursMap,(val, key)=>{...})
```

Beking0912 replied:
Thanks! I've tried to use `_.forIn`, but it returns the original object instead of the content I'm returning, so I used `_.map` in this case.

alisman commented on src/tours/Tour/index.tsx:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * If the url contains a query param 'webtour', set the localStorage
+     */
+    const urlStr = window.location.href.split('?')[1];
```
we should obtain url params from routingStore in calling component.  that already has url parsed.

alisman commented on src/tours/Tour/index.tsx:
```diff
@@ -0,0 +1,140 @@
+    const [startAt, setStartAt] = useState<number>(0);
+    const endTour = () => setIsOpen(false);
+    const endTourWithBtn = (e: any) => {
+        e.preventDefault();
```
are you sure you need these?

Beking0912 replied:
I think so, `e.preventDefault(); e.stopPropagation();` is required in the click event here. 
Without it, it would report `TypeError: Cannot read properties of null (reading 'getBoundingClientRect')` when click 'Skip All' button on the second step of virtual study tour.

alisman commented on src/tours/Tour/index.tsx:
```diff
@@ -0,0 +1,140 @@
+import './styles.scss';
+
+export const setTourLocalStorage = (id: string, value: string) => {
+    localStorage.setItem('web-tour', id);
```
use constant for "web-tour" so that we can use it again for retrieval down below.

alisman commented on src/tours/Steps/index.tsx:
```diff
@@ -0,0 +1,9 @@
+import getVirtualStudySteps, { virtualStudyId } from './VirtualStudy';
```
i'm curious what importing and then exporting from this index file achieves.  i know we do this in packages so that you can import from the package but what here?

Beking0912 replied:
I want to organize the import and export of methods and variables related to steps all within a single file. 
This is just my personal preference, and deleting this file is entirely feasible. 
How do you think about?

alisman replied:
i see.  that makes sense.

alisman commented:
@Beking0912 let me know if you want to discuss some of my comments.  most of these are minor things.  nice work!

alisman commented on src/pages/groupComparison/groupSelector/GroupSelector.tsx:
```diff
@@ -208,6 +208,7 @@ export default class GroupSelector extends React.Component<
                 );
                 return (
                     <div
+                        id="single-study-group-comparison-groups"
```
you should use 
```
data-tour=XXXX
```

alisman replied:
then for the selector it would be 
```
[data-tour=XXXX]
```

Beking0912 replied:
I'm thinking about adding a `web-tour` attribute on the target element, just like `data-test` in the project, such as:
- `<div web-tour="my-first-step" />`
- `selector: [web-tour="my-first-step"]`

How do you think about it, do you have better idea?

Beking0912 replied:
Good idea!

Beking0912 commented on src/tours/Steps/VirtualStudy.tsx:
```diff
@@ -0,0 +1,330 @@
+            <div className="step">
+                <p className="title">Click on the Share button</p>
+                <p>1. Enter a name for your virtual study (optional).</p>
+                2. Text box pre-filled with a description of the studies
```
I believe that whether to include `<p>` here during frontend rendering doesn't make a difference in this context. 
However, for the sake of code standardization, I will add it. Thanks!

Beking0912 commented:
Thanks for your comments, @inodb @alisman ! I've make many updates on my code, take a look whenever you can, and feel free to give me any more suggestions for improvements!
In addition, there are two more related pull requests that require your attention：
- End-to-end testing for Interactive Web Tour PR: https://github.com/cBioPortal/cbioportal-frontend/pull/4700
- Add `show_web_tours` for Interactive Web Tour PR: https://github.com/cBioPortal/cbioportal/pull/10333

Thanks again for your time!
